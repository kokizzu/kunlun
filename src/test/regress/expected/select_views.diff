101a102,103
> psql:sql/select_views.sql:84: NOTICE:  f_leak => beafsteak
> psql:sql/select_views.sql:84: NOTICE:  f_leak => hamburger
108,113c110,114
<                                                                                    QUERY PLAN                                                                                    
< ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
<  RemoteScan on customer
<    Filter: f_leak(passwd)
<    Shard: 2      Remote SQL: select  customer.cid, customer.name, customer.tel, customer.passwd from regress_$$_public.customer customer where (customer.name = 'regress_alice')
< (3 rows)
---
>                           QUERY PLAN                          
> --------------------------------------------------------------
>  Seq Scan on customer
>    Filter: (f_leak(passwd) AND (name = (CURRENT_USER)::text))
> (2 rows)
123,124c124,125
<                                                                                        QUERY PLAN                                                                                        
< -----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
---
>                   QUERY PLAN                   
> -----------------------------------------------
127,128c128,129
<    ->  RemoteScan on customer
<          Shard: 2        Remote SQL: select  customer.cid, customer.name, customer.tel, customer.passwd from regress_$$_public.customer customer where (customer.name = 'regress_alice')
---
>    ->  Seq Scan on customer
>          Filter: (name = (CURRENT_USER)::text)
138a140,143
> psql:sql/select_views.sql:95: NOTICE:  f_leak => passwd
> psql:sql/select_views.sql:95: NOTICE:  f_leak => beafsteak
> psql:sql/select_views.sql:95: NOTICE:  f_leak => passwd
> psql:sql/select_views.sql:95: NOTICE:  f_leak => hamburger
146,151c151,155
<                                                                                    QUERY PLAN                                                                                    
< ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
<  RemoteScan on customer
<    Filter: (f_leak('passwd'::text) AND f_leak(passwd))
<    Shard: 2      Remote SQL: select  customer.cid, customer.name, customer.tel, customer.passwd from regress_$$_public.customer customer where (customer.name = 'regress_alice')
< (3 rows)
---
>                                        QUERY PLAN                                        
> -----------------------------------------------------------------------------------------
>  Seq Scan on customer
>    Filter: (f_leak('passwd'::text) AND f_leak(passwd) AND (name = (CURRENT_USER)::text))
> (2 rows)
156a161,162
> psql:sql/select_views.sql:100: NOTICE:  f_leak => passwd
> psql:sql/select_views.sql:100: NOTICE:  f_leak => passwd
164,165c170,171
<                                                                                        QUERY PLAN                                                                                        
< -----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
---
>                                  QUERY PLAN                                 
> ----------------------------------------------------------------------------
168,171c174,176
<    ->  RemoteScan on customer
<          Filter: f_leak('passwd'::text)
<          Shard: 2        Remote SQL: select  customer.cid, customer.name, customer.tel, customer.passwd from regress_$$_public.customer customer where (customer.name = 'regress_alice')
< (5 rows)
---
>    ->  Seq Scan on customer
>          Filter: (f_leak('passwd'::text) AND (name = (CURRENT_USER)::text))
> (4 rows)
188,194c193,197
<                                                                   QUERY PLAN                                                                   
< -----------------------------------------------------------------------------------------------------------------------------------------------
<  Nested Loop
<    Join Filter: (l.cid = r.cid)
<    ->  RemoteScan on customer l
<          Shard: 2        Remote SQL: select  l.cid, l.name, l.tel, l.passwd from regress_$$_public.customer l where (l.name = 'regress_alice')
<    ->  RemoteScan on credit_card r
---
>                      QUERY PLAN                      
> -----------------------------------------------------
>  Hash Join
>    Hash Cond: (r.cid = l.cid)
>    ->  Seq Scan on credit_card r
196c199,201
<          Shard: 2        Remote SQL: select  r.cid, r.cnum, r.climit from regress_$$_public.credit_card r
---
>    ->  Hash
>          ->  Seq Scan on customer l
>                Filter: (name = (CURRENT_USER)::text)
207,208c212,213
<                                                                                                            QUERY PLAN                                                                                                            
< ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
---
>                         QUERY PLAN                         
> -----------------------------------------------------------
211,214c216,222
<    ->  RemoteJoin Join
<          Join Filter: ((l.cid = r.cid) AND (l.name = (CURRENT_USER)::text))
<          Shard: 2        Remote SQL: select  l.cid, l.name, l.tel, l.passwd, r.cnum, r.climit from  (  regress_$$_public.customer l  JOIN  regress_$$_public.credit_card r   ON (l.cid = r.cid) AND (l.name = 'regress_alice') )
< (5 rows)
---
>    ->  Hash Join
>          Hash Cond: (r.cid = l.cid)
>          ->  Seq Scan on credit_card r
>          ->  Hash
>                ->  Seq Scan on customer l
>                      Filter: (name = (CURRENT_USER)::text)
> (8 rows)
233,234c241,242
<                                                                                                            QUERY PLAN                                                                                                            
< ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
---
>                                   QUERY PLAN                                  
> ------------------------------------------------------------------------------
237,244c245,256
<    ->  Subquery Scan on l
<          Filter: f_leak(l.cnum)
<          ->  RemoteJoin Join
<                Join Filter: ((l_1.cid = r_1.cid) AND (l_1.name = (CURRENT_USER)::text))
<                Shard: 2  Remote SQL: select  l.cid, l.name, l.tel, l.passwd, r.cnum, r.climit from  (  regress_$$_public.customer l  JOIN  regress_$$_public.credit_card r   ON (l.cid = r.cid) AND (l.name = 'regress_alice') )
<    ->  RemoteScan on credit_usage r
<          Shard: 1        Remote SQL: select  r.cid, r.ymd, r.usage1 from regress_$$_public.credit_usage r where (r.ymd >= '2011-10-01') AND (r.ymd < '2011-11-01')
< (9 rows)
---
>    ->  Seq Scan on credit_usage r
>          Filter: ((ymd >= '2011-10-01'::date) AND (ymd < '2011-11-01'::date))
>    ->  Materialize
>          ->  Subquery Scan on l
>                Filter: f_leak(l.cnum)
>                ->  Hash Join
>                      Hash Cond: (r_1.cid = l_1.cid)
>                      ->  Seq Scan on credit_card r_1
>                      ->  Hash
>                            ->  Seq Scan on customer l_1
>                                  Filter: (name = (CURRENT_USER)::text)
> (13 rows)
260,261c272,273
<                                                                                                            QUERY PLAN                                                                                                            
< ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
---
>                                      QUERY PLAN                                     
> ------------------------------------------------------------------------------------
266,271c278,287
<          ->  RemoteJoin Join
<                Join Filter: ((l.cid = r_1.cid) AND (l.name = (CURRENT_USER)::text))
<                Shard: 2  Remote SQL: select  l.cid, l.name, l.tel, l.passwd, r.cnum, r.climit from  (  regress_$$_public.customer l  JOIN  regress_$$_public.credit_card r   ON (l.cid = r.cid) AND (l.name = 'regress_alice') )
<          ->  RemoteScan on credit_usage r
<                Shard: 1  Remote SQL: select  r.cid, r.ymd, r.usage1 from regress_$$_public.credit_usage r where (r.ymd >= '2011-10-01') AND (r.ymd < '2011-11-01')
< (9 rows)
---
>          ->  Seq Scan on credit_usage r
>                Filter: ((ymd >= '2011-10-01'::date) AND (ymd < '2011-11-01'::date))
>          ->  Materialize
>                ->  Hash Join
>                      Hash Cond: (r_1.cid = l.cid)
>                      ->  Seq Scan on credit_card r_1
>                      ->  Hash
>                            ->  Seq Scan on customer l
>                                  Filter: (name = (CURRENT_USER)::text)
> (13 rows)
282a299,300
> psql:sql/select_views.sql:136: NOTICE:  f_leak => beafsteak
> psql:sql/select_views.sql:136: NOTICE:  f_leak => hamburger
302a321,322
> psql:sql/select_views.sql:142: NOTICE:  f_leak => beafsteak
> psql:sql/select_views.sql:142: NOTICE:  f_leak => hamburger
