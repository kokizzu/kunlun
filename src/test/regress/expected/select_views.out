--
-- Test for Leaky view scenario
--
CREATE ROLE regress_alice;
CREATE ROLE
CREATE FUNCTION f_leak (text)
       RETURNS bool LANGUAGE 'plpgsql' COST 0.0000001
       AS 'BEGIN RAISE NOTICE ''f_leak => %'', $1; RETURN true; END';
CREATE FUNCTION
DROP TABLE if exists customer cascade;
psql:sql/select_views.sql:10: NOTICE:  table "customer" does not exist, skipping
DROP TABLE
CREATE TABLE customer (
       cid      int primary key,
       name     text not null,
       tel      text,
       passwd	text
);
CREATE TABLE
DROP TABLE if exists credit_card cascade;
psql:sql/select_views.sql:18: NOTICE:  table "credit_card" does not exist, skipping
DROP TABLE
CREATE TABLE credit_card (
       cid      int,
       cnum     text,
       climit   int
);
CREATE TABLE
DROP TABLE if exists credit_usage cascade;
psql:sql/select_views.sql:25: NOTICE:  table "credit_usage" does not exist, skipping
DROP TABLE
CREATE TABLE credit_usage (
       cid      int,
       ymd      date,
       usage1    int
);
CREATE TABLE
INSERT INTO customer
       VALUES (101, 'regress_alice', '+81-12-3456-7890', 'passwd123'),
              (102, 'regress_bob',   '+01-234-567-8901', 'beafsteak'),
              (103, 'regress_eve',   '+49-8765-43210',   'hamburger');
INSERT 0 3
INSERT INTO credit_card
       VALUES (101, '1111-2222-3333-4444', 4000),
              (102, '5555-6666-7777-8888', 3000),
              (103, '9801-2345-6789-0123', 2000);
INSERT 0 3
INSERT INTO credit_usage
       VALUES (101, '2011-09-15', 120),
	      (101, '2011-10-05',  90),
	      (101, '2011-10-18', 110),
	      (101, '2011-10-21', 200),
	      (101, '2011-11-10',  80),
	      (102, '2011-09-22', 300),
	      (102, '2011-10-12', 120),
	      (102, '2011-10-28', 200),
	      (103, '2011-10-15', 480);
INSERT 0 9
CREATE VIEW my_property_normal AS
       SELECT * FROM customer WHERE name = current_user;
CREATE VIEW
CREATE VIEW my_property_secure WITH (security_barrier) AS
       SELECT * FROM customer WHERE name = current_user;
CREATE VIEW
CREATE VIEW my_credit_card_normal AS
       SELECT * FROM customer l NATURAL JOIN credit_card r
       WHERE l.name = current_user;
CREATE VIEW
CREATE VIEW my_credit_card_secure WITH (security_barrier) AS
       SELECT * FROM customer l NATURAL JOIN credit_card r
       WHERE l.name = current_user;
CREATE VIEW
CREATE VIEW my_credit_card_usage_normal AS
       SELECT * FROM my_credit_card_secure l NATURAL JOIN credit_usage r;
CREATE VIEW
CREATE VIEW my_credit_card_usage_secure WITH (security_barrier) AS
       SELECT * FROM my_credit_card_secure l NATURAL JOIN credit_usage r;
CREATE VIEW
GRANT SELECT ON my_property_normal TO public;
GRANT
GRANT SELECT ON my_property_secure TO public;
GRANT
GRANT SELECT ON my_credit_card_normal TO public;
GRANT
GRANT SELECT ON my_credit_card_secure TO public;
GRANT
GRANT SELECT ON my_credit_card_usage_normal TO public;
GRANT
GRANT SELECT ON my_credit_card_usage_secure TO public;
GRANT
--
-- Run leaky view scenarios
--
SET SESSION AUTHORIZATION regress_alice;
SET
--
-- scenario: if a qualifier with tiny-cost is given, it shall be launched
--           prior to the security policy of the view.
--
SELECT * FROM my_property_normal WHERE f_leak(passwd);
psql:sql/select_views.sql:84: NOTICE:  f_leak => passwd123
 cid |     name      |       tel        |  passwd   
-----+---------------+------------------+-----------
 101 | regress_alice | +81-12-3456-7890 | passwd123
(1 row)

EXPLAIN (COSTS OFF) SELECT * FROM my_property_normal WHERE f_leak(passwd);

SELECT * FROM my_property_secure WHERE f_leak(passwd);
psql:sql/select_views.sql:87: NOTICE:  f_leak => passwd123
 cid |     name      |       tel        |  passwd   
-----+---------------+------------------+-----------
 101 | regress_alice | +81-12-3456-7890 | passwd123
(1 row)

EXPLAIN (COSTS OFF) SELECT * FROM my_property_secure WHERE f_leak(passwd);

--
-- scenario: qualifiers can be pushed down if they contain leaky functions,
--           provided they aren't passed data from inside the view.
--
SELECT * FROM my_property_normal v
		WHERE f_leak('passwd') AND f_leak(passwd);
psql:sql/select_views.sql:95: NOTICE:  f_leak => passwd
psql:sql/select_views.sql:95: NOTICE:  f_leak => passwd123
 cid |     name      |       tel        |  passwd   
-----+---------------+------------------+-----------
 101 | regress_alice | +81-12-3456-7890 | passwd123
(1 row)

EXPLAIN (COSTS OFF) SELECT * FROM my_property_normal v
		WHERE f_leak('passwd') AND f_leak(passwd);

SELECT * FROM my_property_secure v
		WHERE f_leak('passwd') AND f_leak(passwd);
psql:sql/select_views.sql:100: NOTICE:  f_leak => passwd
psql:sql/select_views.sql:100: NOTICE:  f_leak => passwd123
 cid |     name      |       tel        |  passwd   
-----+---------------+------------------+-----------
 101 | regress_alice | +81-12-3456-7890 | passwd123
(1 row)

EXPLAIN (COSTS OFF) SELECT * FROM my_property_secure v
		WHERE f_leak('passwd') AND f_leak(passwd);

--
-- scenario: if a qualifier references only one-side of a particular join-
--           tree, it shall be distributed to the most deep scan plan as
--           possible as we can.
--
SELECT * FROM my_credit_card_normal WHERE f_leak(cnum);
psql:sql/select_views.sql:109: NOTICE:  f_leak => 1111-2222-3333-4444
psql:sql/select_views.sql:109: NOTICE:  f_leak => 5555-6666-7777-8888
psql:sql/select_views.sql:109: NOTICE:  f_leak => 9801-2345-6789-0123
 cid |     name      |       tel        |  passwd   |        cnum         | climit 
-----+---------------+------------------+-----------+---------------------+--------
 101 | regress_alice | +81-12-3456-7890 | passwd123 | 1111-2222-3333-4444 |   4000
(1 row)

EXPLAIN (COSTS OFF) SELECT * FROM my_credit_card_normal WHERE f_leak(cnum);

SELECT * FROM my_credit_card_secure WHERE f_leak(cnum);
psql:sql/select_views.sql:112: NOTICE:  f_leak => 1111-2222-3333-4444
 cid |     name      |       tel        |  passwd   |        cnum         | climit 
-----+---------------+------------------+-----------+---------------------+--------
 101 | regress_alice | +81-12-3456-7890 | passwd123 | 1111-2222-3333-4444 |   4000
(1 row)

EXPLAIN (COSTS OFF) SELECT * FROM my_credit_card_secure WHERE f_leak(cnum);

--
-- scenario: an external qualifier can be pushed-down by in-front-of the
--           views with "security_barrier" attribute, except for operators
--           implemented with leakproof functions.
--
SELECT * FROM my_credit_card_usage_normal
       WHERE f_leak(cnum) AND ymd >= '2011-10-01' AND ymd < '2011-11-01';
psql:sql/select_views.sql:121: NOTICE:  f_leak => 1111-2222-3333-4444
 cid |     name      |       tel        |  passwd   |        cnum         | climit |    ymd     | usage1 
-----+---------------+------------------+-----------+---------------------+--------+------------+--------
 101 | regress_alice | +81-12-3456-7890 | passwd123 | 1111-2222-3333-4444 |   4000 | 2011-10-05 |     90
 101 | regress_alice | +81-12-3456-7890 | passwd123 | 1111-2222-3333-4444 |   4000 | 2011-10-18 |    110
 101 | regress_alice | +81-12-3456-7890 | passwd123 | 1111-2222-3333-4444 |   4000 | 2011-10-21 |    200
(3 rows)

EXPLAIN (COSTS OFF) SELECT * FROM my_credit_card_usage_normal
       WHERE f_leak(cnum) AND ymd >= '2011-10-01' AND ymd < '2011-11-01';

SELECT * FROM my_credit_card_usage_secure
       WHERE f_leak(cnum) AND ymd >= '2011-10-01' AND ymd < '2011-11-01';
psql:sql/select_views.sql:126: NOTICE:  f_leak => 1111-2222-3333-4444
psql:sql/select_views.sql:126: NOTICE:  f_leak => 1111-2222-3333-4444
psql:sql/select_views.sql:126: NOTICE:  f_leak => 1111-2222-3333-4444
 cid |     name      |       tel        |  passwd   |        cnum         | climit |    ymd     | usage1 
-----+---------------+------------------+-----------+---------------------+--------+------------+--------
 101 | regress_alice | +81-12-3456-7890 | passwd123 | 1111-2222-3333-4444 |   4000 | 2011-10-05 |     90
 101 | regress_alice | +81-12-3456-7890 | passwd123 | 1111-2222-3333-4444 |   4000 | 2011-10-18 |    110
 101 | regress_alice | +81-12-3456-7890 | passwd123 | 1111-2222-3333-4444 |   4000 | 2011-10-21 |    200
(3 rows)

EXPLAIN (COSTS OFF) SELECT * FROM my_credit_card_usage_secure
       WHERE f_leak(cnum) AND ymd >= '2011-10-01' AND ymd < '2011-11-01';

--
-- Test for the case when security_barrier gets changed between rewriter
-- and planner stage.
--
PREPARE p1 AS SELECT * FROM my_property_normal WHERE f_leak(passwd);
PREPARE
PREPARE p2 AS SELECT * FROM my_property_secure WHERE f_leak(passwd);
PREPARE
EXECUTE p1;
psql:sql/select_views.sql:136: NOTICE:  f_leak => passwd123
 cid |     name      |       tel        |  passwd   
-----+---------------+------------------+-----------
 101 | regress_alice | +81-12-3456-7890 | passwd123
(1 row)

EXECUTE p2;
psql:sql/select_views.sql:137: NOTICE:  f_leak => passwd123
 cid |     name      |       tel        |  passwd   
-----+---------------+------------------+-----------
 101 | regress_alice | +81-12-3456-7890 | passwd123
(1 row)

RESET SESSION AUTHORIZATION;
RESET
--ALTER VIEW my_property_normal SET (security_barrier=true);
--ALTER VIEW my_property_secure SET (security_barrier=false);
SET SESSION AUTHORIZATION regress_alice;
SET
EXECUTE p1;		-- To be perform as a view with security-barrier
psql:sql/select_views.sql:142: NOTICE:  f_leak => passwd123
 cid |     name      |       tel        |  passwd   
-----+---------------+------------------+-----------
 101 | regress_alice | +81-12-3456-7890 | passwd123
(1 row)

EXECUTE p2;		-- To be perform as a view without security-barrier
psql:sql/select_views.sql:143: NOTICE:  f_leak => passwd123
 cid |     name      |       tel        |  passwd   
-----+---------------+------------------+-----------
 101 | regress_alice | +81-12-3456-7890 | passwd123
(1 row)

-- Cleanup.
RESET SESSION AUTHORIZATION;
RESET
DROP ROLE regress_alice;
DROP ROLE
